cajita/src/Cajita_SparseIndexSpace.hpp

```diff
@@ -0,0 +1,817 @@
+    {
+        return _op_key2ijk( key, tile_i, tile_j, tile_k );
+    }
+
+  private:
+    //! [0] current valid table size, [1] pre-allocated size
+    Kokkos::View<int[2], ExecutionSpace> _tile_table_info;
+    //! label to indicate if the tile is being or has been inserted (0 for
+    //! false, other for true)
+    Kokkos::View<int*, ExecutionSpace> _tile_inserted_label;
+    //! current number of tiles inserted to the hash table
+    Kokkos::Array<int, 3> _block_size;
+    //! default factor, rehash by which when need more capacity of the hash
+    Kokkos::Array<float, 1> _rehash_coeff;
+    //! hash table (tile hash key => tile No)
+    Kokkos::UnorderedMap<key_type, value_type, ExecutionSpace> _tile_table;
+    //! Ops: transfer between tile ijk <=> tile hash key
+    TileID2HashKey<key_type, hash_type> _op_ijk2key;
+    HashKey2TileID<key_type, hash_type> _op_key2ijk;
+};
+
+//---------------------------------------------------------------------------//
+/*!
+  \class BlockIndexSpace
+  \brief Block index space, mapping tile ijks to tile No. through a hash table
+  (Kokkos unordered map) \tparam CBits Bits number (per dimension) neded to
+  index cells inside each tile \tparam CNumPerDim Cell number (per dimension)
+  inside each tile \tparam CNumPerTile Cel number (total) inside each tile
+*/
+template <int CBits, int CNumPerDim, int CNumPerTile>
+class TileIndexSpace
+{
+  public:
+    //! Number of bits (per dimension) needed to index the cells inside a tile
+    static constexpr int cell_bits_per_tile_dim = CBits;
+    //! Number of cells (per dimension) inside each tile
+    static constexpr int cell_num_per_tile_dim = CNumPerDim;
+    //! Number of cells (total) inside each tile
+    static constexpr int cell_num_per_tile = CNumPerTile;
+    //! Indexing dimension, currently only serves 3D representations
+    static constexpr int rank = 3;
+
+    /*!
+      \brief
+    */
+    //! Cell ijk <=> Cell Id
+    template <typename... Coords>
+    KOKKOS_INLINE_FUNCTION static constexpr auto
+    coordToOffset( Coords&&... coords ) -> uint64_t
+    {
+        return fromCoord<Coord2OffsetDim>( std::forward<Coords>( coords )... );
+    }
+
+    /*!
+      \brief
+    */
+    template <typename Key, typename... Coords>
+    KOKKOS_INLINE_FUNCTION static constexpr void
+    offsetToCoord( Key&& key, Coords&... coords )
+    {
+        toCoord<Offset2CoordDim>( std::forward<Key>( key ), coords... );
+    }
+
+  protected:
+    //! Coord  <=> Offset Computations
+    /*!
+      \brief Transfer function: tile ijk to tile key
+    */
+    struct Coord2OffsetDim
+    {
+        template <typename Coord>
+        KOKKOS_INLINE_FUNCTION constexpr auto operator()( int dim_no,
+                                                          Coord&& i )
+            -> uint64_t
+        {
+            uint64_t result = (uint64_t)i;
+            for ( int i = 0; i < dim_no; i++ )
```

Comment:
If we knew that we were always 3D we could hand unroll this loop and get rid of some of these multiply operations.